src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -64,10 +89,143 @@ public override int SaveChanges(IList<IUpdateEntry> entries)
     /// <returns>Task that when resolved contains the number of documents affected during saving changes.</returns>
     public override async Task<int> SaveChangesAsync(IList<IUpdateEntry> entries, CancellationToken cancellationToken = default)
     {
-        var updates = MongoUpdate.CreateAll(GetAllChangedRootEntries(entries));
-        return (int)await _mongoClient.SaveUpdatesAsync(updates, cancellationToken).ConfigureAwait(false);
+        var rootEntries = GetAllChangedRootEntries(entries);
+        var updates = MongoUpdate.CreateAll(rootEntries);
+
+        using var session = await _mongoClient.StartSessionAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+
+        return ShouldStartTransaction(rootEntries.Count)
+            ? await ExecuteInTransactionAsync(c => WriteBatchesAsync(updates, session, c), session, cancellationToken).ConfigureAwait(false)
+            : await WriteBatchesAsync(updates, session, cancellationToken).ConfigureAwait(false);
+    }
+
+    private T ExecuteInTransaction<T>(Func<T> operation, IClientSession session)
+    {
+       var transaction = StartTransaction(session, false);
+
+        T result;
+        try
+        {
+            result = operation();
+        }
+        catch
+        {
+            transaction.Rollback();
+            throw;
+        }
+
+        transaction.Commit();
+
+        return result;
+    }
+
+    private async Task<T> ExecuteInTransactionAsync<T>(Func<CancellationToken, Task<T>> operation, IClientSession session, CancellationToken cancellationToken)
+    {
+        var transaction = StartTransaction(session, true);
+
+        T result;
+        try
+        {
+            result = await operation(cancellationToken).ConfigureAwait(false);
+        }
+        catch
+        {
+            await transaction.RollbackAsync(cancellationToken).ConfigureAwait(false);
+            throw;
+        }
+
+        await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
+        return result;
     }
 
+    private MongoTransaction StartTransaction(IClientSession session, bool async)
```

Comment:
Consider making this a factory method on mongo transaction

Reply:
Done.